File: Strohm/android/strohm-native/src/main/java/dev/strohmnative/viewmodels/SimpleViewModel.kt
package dev.strohmnative.viewmodels

import android.util.Log
import dev.strohmnative.*

open class SimpleViewModel<EntryType>(
    initialData: EntryType,
    propName: PropName,
    propPath: PropPath,
    private val instanceFactory: ConstructableFromDictionary<EntryType>
): ViewModelBase<EntryType>(initialData, propName, propPath) {
    override fun propToData(prop: Prop): EntryType? {
      if (prop.first != propName) { return null; }

      val m = prop.second as? Map<*, *> ?: return null
        val rawData = m.asMapOfKeyType<String>() ?: return null
        var data: EntryType? = instanceFactory.createFromDict(rawData) ?: return null

        Log.d("dev.strohmnative", "Received entry: $data")
        return data
    }
}

File: Strohm/ios/StrohmNative/Sources/StrohmNative/ViewModels/SimpleViewModel.swift
import Foundation

open class SimpleViewModel<EntryType: Decodable>: ViewModelBase<EntryType> {

    public var data: EntryType {
        willSet {
            // https://stackoverflow.com/questions/57615920/published-property-wrapper-not-working-on-subclass-of-observableobject
            self.objectWillChange.send()
        }
    }

    public init(initialData: EntryType, propName: PropName, propPath: PropPath) {
        self.data = initialData
        super.init(propName: propName, propPath: propPath)
    }

    public required init(constant: EntryType) {
        self.data = constant
        super.init()
    }

    override func propToData2(serializedProp: String) -> EntryType? {
        guard let rawData = serializedProp.data(using: .utf8) else {
            return nil
        }

        do {
            let data = try decoder().decode(PropEnvelope<EntryType?>.self, from: rawData)
            let value = data.propValue
            print("Received entry: ", value)
            return value
        }
        catch let e {
            Log.error(String(describing: e))
            return nil
        }
    }

    override func store(data: EntryType) {
        self.data = data
    }

    public static func constant(_ data: EntryType) -> Self {
        return .init(constant: data)
    }
}
